packages/translator-default/src/analyze.js

```diff
@@ -0,0 +1,54 @@
+import { types as t } from "@marko/babel-types";
+import { isNativeTag, getTagDef } from "@marko/babel-utils";
+
+export const staticNodes = new WeakSet();
+
+export const visitor = {
+  MarkoText(path) {
+    staticNodes.add(path.node);
+  },
+  MarkoPlaceholder(path) {
+    const { confident } = path.get("value").evaluate();
+    if (confident && path.node.escape) staticNodes.add(path.node);
+  },
+  MarkoTag: {
+    enter(path) {
+      // check if a for
+      // needed to handle global keys on elements that don't have specific key attributes
+      if (path.node.name.value === "for") path.skip();
```

Comment:
Shouldn't `while` be considered here as well? And maybe some other control flow tags?

Reply:
This is due to implicit keys under loops that break when hoisted. It looks like we use these keys not just top level in the loop. "while" makes sense are there others this applies to? For was the only one in the tests so it was the only one I hit issues with.